src/rgw/services/svc_zone.cc

```diff
@@ -47,157 +47,167 @@ bool RGWSI_Zone::zone_syncs_from(const RGWZone& target_zone, const RGWZone& sour
          sync_modules_svc->get_manager()->supports_data_export(source_zone.tier_type);
 }
 
-int RGWSI_Zone::do_start()
+boost::system::error_code RGWSI_Zone::do_start()
 {
-  int ret = sysobj_svc->start();
-  if (ret < 0) {
+  auto ret = sysobj_svc->start();
+  if (ret) {
     return ret;
   }
 
   assert(sysobj_svc->is_started()); /* if not then there's ordering issue */
 
   ret = rados_svc->start();
-  if (ret < 0) {
+  if (ret) {
     return ret;
   }
   ret = sync_modules_svc->start();
-  if (ret < 0) {
+  if (ret) {
     return ret;
   }
-  ret = realm->init(cct, sysobj_svc);
-  if (ret < 0 && ret != -ENOENT) {
-    ldout(cct, 0) << "failed reading realm info: ret "<< ret << " " << cpp_strerror(-ret) << dendl;
-    return ret;
-  } else if (ret != -ENOENT) {
-    ldout(cct, 20) << "realm  " << realm->get_name() << " " << realm->get_id() << dendl;
-    ret = current_period->init(cct, sysobj_svc, realm->get_id(), realm->get_name());
-    if (ret < 0 && ret != -ENOENT) {
-      ldout(cct, 0) << "failed reading current period info: " << " " << cpp_strerror(-ret) << dendl;
-      return ret;
-    }
-    ldout(cct, 20) << "current period " << current_period->get_id() << dendl;  
-  }
 
-  ret = replace_region_with_zonegroup();
-  if (ret < 0) {
-    lderr(cct) << "failed converting region to zonegroup : ret "<< ret << " " << cpp_strerror(-ret) << dendl;
-    return ret;
-  }
+  auto more_init =
+    [this]() -> boost::system::error_code {
+      auto ret = ceph::to_error_code(realm->init(cct, sysobj_svc));
+      if (ret && ret != boost::system::errc::no_such_file_or_directory) {
+        ldout(cct, 0) << "failed reading realm info: ret "<< ret << " " << dendl;
+        return ret;
+      } else if (ret == boost::system::errc::no_such_file_or_directory) {
```

Comment:
should be `ret !=` here

p.s. the `more_init` lambda doesn't appear to be needed anymore

Reply:
cherry-picked

Reply:
more_init was a bt of scaffolding, yes. Thank you.